providers/implementations/keymgmt/slh_dsa_kmgmt.c

```diff
@@ -0,0 +1,436 @@
+        return ossl_param_build_set_octet_string(tmpl, NULL,
+                                                 OSSL_PKEY_PARAM_PRIV_KEY,
+                                                 ossl_slh_dsa_key_get_priv(key),
+                                                 ossl_slh_dsa_key_get_priv_len(key));
+    /* Otherwise write out the public key element */
+    return ossl_param_build_set_octet_string(tmpl, NULL,
+                                             OSSL_PKEY_PARAM_PUB_KEY,
+                                             ossl_slh_dsa_key_get_pub(key),
+                                             ossl_slh_dsa_key_get_pub_len(key));
+}
+
+static int slh_dsa_get_params(void *keydata, OSSL_PARAM params[])
+{
+    SLH_DSA_KEY *key = keydata;
+    OSSL_PARAM *p;
+    const uint8_t *pub, *priv;
+
+    if ((p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_BITS)) != NULL
+            && !OSSL_PARAM_set_int(p, 8 * ossl_slh_dsa_key_get_pub_len(key)))
+        return 0;
+    if ((p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_SECURITY_BITS)) != NULL
+            && !OSSL_PARAM_set_int(p, 8 * ossl_slh_dsa_key_get_n(key)))
+        return 0;
+    if ((p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_MAX_SIZE)) != NULL
+            && !OSSL_PARAM_set_int(p, ossl_slh_dsa_key_get_sig_len(key)))
+        return 0;
+
+    pub = ossl_slh_dsa_key_get_pub(key);
+    priv = ossl_slh_dsa_key_get_priv(key);
+
+    /* This just gets the private elements */
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_PRIV_KEY);
+    if (p != NULL) {
+        if (priv == NULL
+                || !OSSL_PARAM_set_octet_string(p, priv,
+                                                ossl_slh_dsa_key_get_priv_len(key) / 2))
+            return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_PUB_KEY);
+    if (p != NULL) {
+        if (pub == NULL
+                || !OSSL_PARAM_set_octet_string(p, pub,
+                                                ossl_slh_dsa_key_get_pub_len(key)))
+            return 0;
+    }
+    return 1;
+}
+
+static int slh_dsa_export(void *keydata, int selection, OSSL_CALLBACK *param_cb,
+                          void *cbarg)
+{
+    SLH_DSA_KEY *key = keydata;
+    OSSL_PARAM_BLD *tmpl;
+    OSSL_PARAM *params = NULL;
+    int ret = 0, include_private;
+
+    if (!ossl_prov_is_running() || key == NULL)
+        return 0;
+
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0)
+        return 0;
+
+    tmpl = OSSL_PARAM_BLD_new();
+    if (tmpl == NULL)
+        return 0;
+
+    include_private = ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0);
+    if (!key_to_params(key, tmpl, include_private))
+        goto err;
+
+    params = OSSL_PARAM_BLD_to_param(tmpl);
+    if (params == NULL)
+        goto err;
+
+    ret = param_cb(params, cbarg);
+    OSSL_PARAM_free(params);
+err:
+    OSSL_PARAM_BLD_free(tmpl);
+    return ret;
+}
+
+static void *slh_dsa_load(const void *reference, size_t reference_sz)
```

Comment:
Perhaps `#ifndef FIPS_MODULE` for this function.

Reply:
If we decide to do that it should be applied at a later date to all keymanagement files.